File: frontend/src/renderer/lib/block-viewport.ts
export const PINNED_SLACK_PX = 24;

export const ESTIMATED_BLOCK_HEIGHT = 96;

export const BLOCK_OVERSCAN = 6;

export const VIRTUALIZATION_THRESHOLD = 100;

export function virtualizationThreshold(): number {
	const override = (globalThis as typeof globalThis & {
		__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD?: unknown;
	}).__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD;
	return Number.isFinite(override) ? (override as number) : VIRTUALIZATION_THRESHOLD;
}

export type TopItem = { index: number; start: number; size: number };

export function topItemFor(items: readonly TopItem[], scrollTop: number): TopItem | undefined {
	return items.find((item) => item.start <= scrollTop && item.start + item.size > scrollTop);
}

export function headerSticks(blockHeight: number, viewportHeight: number): boolean {
	return viewportHeight > 0 && blockHeight <= viewportHeight;
}

export function isPinned(scrollTop: number, totalSize: number, viewportHeight: number): boolean {
	return scrollTop >= totalSize - viewportHeight - PINNED_SLACK_PX;
}

export function nextBoundary(current: number | undefined, count: number): number | undefined {
	if (count === 0) return undefined;
	if (current === undefined) return 0;
	const next = current + 1;
	return next >= count ? undefined : next;
}

export function previousBoundary(current: number | undefined, count: number): number | undefined {
	if (count === 0 || current === undefined) return undefined;
	const previous = current - 1;
	return previous < 0 ? undefined : previous;
}

export function previousTarget(
	top: TopItem | undefined,
	scrollTop: number,
	count: number,
): number | undefined {
	if (top === undefined) return undefined;
	if (scrollTop - top.start > 1) return top.index;
	return previousBoundary(top.index, count);
}

File: packages/mobile/lib/feature/blocks/logic/block_viewport.dart
import 'package:operator_mobile/feature/blocks/logic/session_block.dart';

const double kTailSlack = 24;

const int kMaxFollowHops = 20;

sealed class BlockViewport {
  static int pivotIndex(List<SessionBlock> blocks, int? pivotSeq) {
    if (pivotSeq == null) return 0;
    for (var index = 0; index < blocks.length; index++) {
      if (blocks[index].firstSeq >= pivotSeq) return index;
    }
    return blocks.length;
  }

  static bool isPinned(double pixels, double maxScrollExtent) =>
      pixels >= maxScrollExtent - kTailSlack;

  static bool headerSticks(double blockHeight, double viewportHeight) =>
      blockHeight <= viewportHeight;

  static int? nextBoundary(int? current, int count) {
    if (count == 0) return null;
    if (current == null) return 0;
    final next = current + 1;
    return next >= count ? null : next;
  }

  static int? previousBoundary(int? current, int count) {
    if (count == 0 || current == null) return null;
    final previous = current - 1;
    return previous < 0 ? null : previous;
  }
}
